### PR TITLE
Change to only allow a single output type

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

File: docs/src/main/asciidoc/building-native-image-guide.adoc [hunk 1]
```diff
@@ -352,6 +341,14 @@ If you need SSL support in your native executable, you can easily include the ne
 Please see link:native-and-ssl-guide.html#working-with-containers[our Using SSL With Native Executables guide] for more information.
 ====
 
+== Configuring the Native Image
+
+There are a lot of different configuration options that can affect how the native image is generated.
+These are provided in `application.properties` the same as any other config property.
+
+
+These options can be seen in the link:all-config.html[All Config Documentation] under the prefix `quarkus.native.`.
```
Comment (gsmet) [inline docs/src/main/asciidoc/building-native-image-guide.adoc]: I would rewrite this paragraph and include the properties directly with:
```
include::{generated-dir}/config/quarkus-core-native-config.adoc[opts=optional]
```

**Review by gsmet — CHANGES_REQUESTED** [review]
I think this is mostly good to go but here are a few fixes and suggestions.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/PackageTypeVerificationBuildStep.java [hunk 2]
```diff
@@ -16,17 +17,15 @@
...
     @BuildStep
     List<PackageTypeBuildItem> builtins() {
-        return Arrays.asList(new PackageTypeBuildItem(PackageConfig.NATIVE), new PackageTypeBuildItem(PackageConfig.THIN_JAR),
-                new PackageTypeBuildItem(PackageConfig.UBER_JAR));
+        return Arrays.asList(new PackageTypeBuildItem(PackageConfig.NATIVE), new PackageTypeBuildItem(PackageConfig.JAR));
     }
 
     @BuildStep
-    void verify(List<PackageTypeBuildItem> items, PackageConfig config) {
+    ServiceStartBuildItem verify(List<PackageTypeBuildItem> items, PackageConfig config) {
```
[the remark this answers is not in the document]
Comment (stuartwdouglas) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/PackageTypeVerificationBuildStep.java]: So it is run at all

File: docs/src/main/asciidoc/building-native-image-guide.adoc [hunk 3]
```diff
@@ -101,36 +101,21 @@ If you have generated the application from the previous tutorial, you can find i
...
-                                <enableHttpUrlHandler>true</enableHttpUrlHandler>
-                            </configuration>
-                        </execution>
-                    </executions>
-                </plugin>
-            </plugins>
-        </build>
+        <properties>
+            <quarkus.package.type>native</quarkus.package.type>
+        </properties>
```
[the remark this answers is not in the document]
Comment (stuartwdouglas) [reply docs/src/main/asciidoc/building-native-image-guide.adoc]: Yes, the current way still works, so existing projects will work as normal.

File: docs/src/main/asciidoc/building-native-image-guide.adoc [hunk 4]
```diff
@@ -101,36 +101,21 @@ If you have generated the application from the previous tutorial, you can find i
...
     </profile>
 </profiles>
 ----
 
 [TIP]
 ====
-You can provide additional options for `native-image` command using `<additionalBuildArgs>foo,bar</additionalBuildArgs>`
-configuration element. Provided options have to be `,` separated.
+You can provide additional options for `native-image` command using `quarkus.native.additional-buildA-args=foo,bar` in `application.properties`. Provided options have to be `,` separated.
 ====
```
[the remark this answers is not in the document]
Comment (stuartwdouglas) [reply docs/src/main/asciidoc/building-native-image-guide.adoc]: The pom.xml way still works, and all the properties have the same name. We should definitely document this, but it is not something that should cause any breakage.

**Comment by dmlloyd** [comment]
This is a bit of a major design change, did I miss the list discussion about this?

**Comment by stuartwdouglas** [comment]
The problem is that without this change we don't really get much extra flexibility from doing native in a build step (and eventually in an extension). You can produce extra stuff that goes into the image, but that is it.

**Comment by stuartwdouglas** [comment]
There are also quite a few other problems with the current build step approach:
- Anything that wants to consume the output (e.g. to zip it up into a deployment package) must pick the type of jar they want to consume. If the user has selected a different type both are generated.
- If there is the possibility of generating both we need different names for the uberjar and thin jar in the output dir.

If this PR is a problem then I would rather revert the change to building in build steps and worry about this post 1.0 than stick with the current approach of allowing multiple outputs, as I think it is very limiting and not something we want to get stuck supporting.

**Comment by dmlloyd** [comment]
There are other ways to get the same kinds of additional flexibility (which we'd want to enumerate anyway).  I think we should have a discussion about whether we want to go this route.

In principle I *like* it because it's simpler - however it's also pretty different from what we had discussed previously from a usability standpoint.  We basically have to be OK with telling the user that they need multiple modules or profiles or something to build multiple output types.

I'm OK with revisiting it later though if nobody else objects.

**Comment by stuartwdouglas** [comment]
You could do it by having some way of filtering particular build items so they only apply to particular output types, but it is much more complex.

Technically an extension could actually force the generation of multiple output types if required, e.g. by consuming both the JarBuildItem and NativeImageBuildItem. I can't really think of a valid use case where you would want both Uber and Thin jars though, at the moment you can't do it and AKAIK nobody has asked for it.

**Comment by stuartwdouglas** [comment]
@dmlloyd are you ok with this now after our conversation the other day?

**Comment by dmlloyd** [comment]
Yes I'm OK with it.  I think that even if we introduce the multi-augment strategy, this change would be a reasonable basis to start with.

**Review by dmlloyd — APPROVED** [review]

**Review by gsmet — APPROVED** [review]

**Pull request merged**